lib/features/push_notification/presentation/controller/web_socket_controller.dart

```diff
@@ -0,0 +1,139 @@
+    if (success is WebSocketConnectionSuccess) {
+      _handleWebSocketConnectionSuccess(success);
+    }
+  }
+
+  @override
+  void handleErrorViewState(Object error, StackTrace stackTrace) {
+    super.handleErrorViewState(error, stackTrace);
+    handleFailureViewState(WebSocketConnectionFailed(exception: null));
+  }
+
+  @override
+  void initialize({AccountId? accountId, Session? session}) {
+    super.initialize(accountId: accountId, session: session);
+
+    _connectWebSocket(accountId, session);
+  }
+
+  void _connectWebSocket(AccountId? accountId, Session? session) {
+    _connectWebSocketInteractor = getBinding<ConnectWebSocketInteractor>();
+    if (_connectWebSocketInteractor == null || accountId == null || session == null) {
+      return;
+    }
+
+    consumeState(_connectWebSocketInteractor!.execute(session, accountId));
+  }
+  
+  void _handleWebSocketConnectionSuccess(WebSocketConnectionSuccess success) {
+    log('WebSocketController::_handleWebSocketConnectionSuccess(): $success');
+    _retryRemained = 3;
+    _webSocketChannel = success.webSocketChannel;
+    _enableWebSocketPush();
+    _pingWebSocket();
+    _listenToWebSocket();
+  }
+
+  void _handleWebSocketConnectionRetry() {
+    if (_retryRemained > 0) {
+      _retryRemained--;
+      _connectWebSocket(accountId, session);
+    }
+  }
+
+  void _enableWebSocketPush() {
+    _webSocketChannel?.sink.add(jsonEncode(WebSocketPushEnableRequest.toJson(
+      dataTypes: [TypeName.emailType, TypeName.mailboxType]
+    )));
+  }
+
+  void _pingWebSocket() {
+    _webSocketPingTimer = Timer.periodic(const Duration(seconds: 10), (_) {
+      _webSocketChannel?.sink.add(jsonEncode(WebSocketEchoRequest.toJson()));
+    });
+  }
+
+  void _listenToWebSocket() {
+    _webSocketSubscription = _webSocketChannel?.stream.listen(
+      (data) {
+        log('WebSocketController::_listenToWebSocket(): $data');
+        if (session == null || accountId == null) return;
+        if (data is String) {
+          data = jsonDecode(data);
+        }
+
+        try {
+          final stateChange = StateChange.fromJson(data);
+          final mapTypeState = stateChange.getMapTypeState(accountId!);
+          mappingTypeStateToAction(
+            mapTypeState,
+            accountId!,
+            emailChangeListener: EmailChangeListener.instance,
+            mailboxChangeListener: MailboxChangeListener.instance,
+            session!.username,
+            session: session);
+        } catch (e) {
+          logError('WebSocketController::_listenToWebSocket(): Data is not StateChange');
+        }
+      },
+      cancelOnError: true,
+      onError: (error) {
+        logError('WebSocketController::_listenToWebSocket():Error: $error');
+        handleFailureViewState(WebSocketConnectionFailed(exception: error));
+      },
+      onDone: () {
+        log('WebSocketController::_listenToWebSocket():onDone');
+        handleFailureViewState(WebSocketConnectionFailed(exception: null));
```

Comment:
Should not call failure at here. We should call retry websocket.

Reply:
Done